packages/control-hub/src/discovery.ts

```diff
@@ -0,0 +1,81 @@
+import { ControlHubInternal } from "./internal/ControlHub.js";
+import { Adb, DeviceClient } from "@u4/adbkit";
+import getPort from "get-port";
+import { ControlHub } from "./ControlHub.js";
+
+export async function openConnectedControlHub(): Promise<ControlHub | undefined> {
+    try {
+        return await createWiFiControlHub();
+    } catch (e: any) {
+        return undefined;
+    }
+}
+
+export async function openUsbControlHubs(): Promise<ControlHub[]> {
+    let adbClient = Adb.createClient();
+    let controlHubs: ControlHub[] = [];
+
+    let devices = await adbClient.listDevices();
+    for (const device of devices) {
+        let deviceClient = device.getClient();
+        let isHub = await isControlHub(deviceClient);
+        if (isHub) {
+            let port = await configureHubTcp(deviceClient);
+            let serialNumber = device.id;
+
+            let hub = new ControlHubInternal(serialNumber);
+            await hub.open("127.0.0.1", (port + 1).toString());
+            controlHubs.push(hub);
+        }
+    }
+
+    return controlHubs;
+}
+
+async function configureHubTcp(deviceClient: DeviceClient): Promise<number> {
+    let port = await findAdjacentPorts();
+    await deviceClient.forward(`tcp:${port}`, `tcp:8080`);
+    await deviceClient.forward(`tcp:${port + 1}`, `tcp:8081`);
+
+    return port;
+}
+
+async function isControlHub(deviceClient: DeviceClient): Promise<boolean> {
+    let serialasusb = await deviceClient.execOut(
+        "getprop persist.ftcandroid.serialasusb",
+        "utf8",
+    );
```

Comment:
This needs to catch whatever error is thrown for devices that don't have the `shell` command enabled for ADB.